Vox importer rewrite. New features and fixes

ElectronicBlueberry opened the pull request:
A rewrite on the vox file importer, with minor additions, tweaks and fixes to the other importers, and voxel objects.
Importer partially based on: https://github.com/jpaver/opengametools/blob/master/src/ogt_vox.h

The main new feature is the ability to fully import Magica Voxel files as scenes, while keeping all groups and objects separate. This allows for a faster workflow with Magica Voxel, skipping previously required import steps, when importing complex vox files.

This PR also fixes a missing sRGB flag on the default material, wrong forward orientation for imported voxel models, and incompatibility with newer vox files. It also introduces a new option for a per-model voxel size, and allows for more control over centering on import.

**sRGB Color, and Vox compatibility example:**

| Old | New |
| ---------- | ---------- |
| ![ColorAndFacingOld](https://user-images.githubusercontent.com/44241786/126084358-ddbc5862-9761-4f5f-a648-f5a547455205.png) | ![ColorAndFacingNew](https://user-images.githubusercontent.com/44241786/126084359-92e86afe-c85e-4e66-b4c7-c12f0b170c21.png) |
| ![CharOld](https://user-images.githubusercontent.com/44241786/126084591-507a9805-f6bf-4580-9028-01b5649449bd.png) | ![CharNew](https://user-images.githubusercontent.com/44241786/126084594-ae6ed570-4e9e-483d-a914-32069a67981f.png) |

As seen in the pictures, the forward direction is also changed. Previously the vox importer kept Magica Voxels x axis direction, but as a result models faced backwards by default. The new importer flips the x-axis, keeping the forward direction from Magica Voxel aligned to Godots -z forward direction.

**New import options, and importer:**

| import options | new importer | imported vox file, with new importer |
| ---------- | ---------- | ---------- |
| ![OptionsNew](https://user-images.githubusercontent.com/44241786/126084652-f7bd8b71-be9a-4465-8ed1-93e17eda8e95.png) | ![SceneImporter](https://user-images.githubusercontent.com/44241786/126084673-3d304654-0f2b-4aae-9130-cbecda8a6691.png) | ![ImportedScene](https://user-images.githubusercontent.com/44241786/126084676-0cd3cee1-f503-4792-bfbf-84e324b277e9.png) |

The "Voxel Size" is now no longer global, and can be set per object. This works with voxel meshes, as well as voxel objects, and the voxel editor. A new editor option has been added, to re-scale the voxels while editing.

The new Magica Voxel Scene importer imports a vox file as seperate objects, keeping all included groups, shapes, and their respective transforms and names. Shapes are imported as Mesh Instances, and Groups as Spatials. This feature was the key reason for starting this PR, as it drastically speeds up the workflow in my current project.


Note: In it's current form this PR is a WIP. While all features I set out to add are present, it has not yet been tested thoroughly enough to be considered production ready. I would appreciate testing and feedback.

ClarkThyLord commented on addons/voxel-core/classes/readers/vox.gd:
```diff
@@ -26,61 +24,368 @@ const magicavoxel_default_palette := [
+		var transformed_voxels := {}
+		var transform: Transform = tree.transform
+		
+		# magica voxel positions are calculated from the models centre
```
Spelling mistake:
> centre -> center

ClarkThyLord commented on addons/voxel-core/engine/importers/meshes.gd:
```diff
@@ -106,3 +171,14 @@ func import(source_file : String, save_path : String, options : Dictionary, r_pl
+
+# gets the fraction needed to offset the origin
+# returns -1, for "no change"
+func _get_frac_for_origin(option_val: int) -> float:
```
I see that some functions are repeated in these EditorImportPlugin files, if we do make a parent class for these EditorImportPlugins we should move common functions to there to. Or maybe move them to Voxel class if they're useful for other things?

ClarkThyLord commented on addons/voxel-core/engine/importers/meshes.gd:
```diff
@@ -5,7 +5,12 @@ extends EditorImportPlugin
 
 ## Enums
-enum Presets { DEFAULT }
+enum Presets {
```
I see that these presets are repeated in multiple EditorImportPlugins, maybe it would be best to make a parent class for similar EditorImportPlugins, I feel it's best to keep the code DRY.

ClarkThyLord commented on addons/voxel-core/engine/importers/vox_scenes.gd:
```diff
@@ -0,0 +1,300 @@
+	return error
+
+
+# recursivly builds scene from dict tree
```
Spelling mistake:
> recursivly -> recursively

ClarkThyLord commented on addons/voxel-core/classes/readers/vox.gd:
```diff
@@ -26,61 +24,368 @@ const magicavoxel_default_palette := [
+				var _layer_id := file.get_32()
+				var _frames := file.get_32()
+				
+				# get the ndoes transforms
```
Spilling mistake:
> ndoes -> nodes

ClarkThyLord reviewed (changes requested):
First of, thanks, this is truly amazing work, as you said this is definitely something much needed to have a faster and more fluid workflow between MagicaVoxel and Godot.

Now, I've done a surface review of the code and everything seems to look good. Only needs a bit cleaning up and formatting, but that can be done quickly so no worries. I'll take a closer look and test it more thoroughly this weekend with some practical cases to make sure everything is ready to merge. 👍

I also see that besides the MagicaVoxel importer rework, you've also taken the liberty of having the `voxel_size` be defined per VoxelObject. This is something I had originally contemplated when implementing the plugin, that is: having the voxel's size be static for all VoxelObjects or have every VoxelObject define their own voxel size. I ended up making it static since I couldn't think of many cases where having each VoxelObject define their own voxel size would be useful, since you could always scale the VoxelObject, and therefore its voxels, as you could with any other MeshInstance. However, if you feel that there is a need for this change then I'm more than happy to accept these changes.

Thanks again, amazing work, looking forward to trying it out! 🙌

ElectronicBlueberry commented:
Thanks for the quick response and surface review!

I resolved the requested changes, and added a importer base class, removing the duplicate code.
For the centering to work consistently across importers, I changed the voxel_object centering function to assume that 0 represents the lowest relative position in the mesh, and 1 the highest. Since this might be a breaking change for some projects, I could alternatively change the scene importer to swap the relative coordinates around.

As for the per object voxel size: It's definitely not a must have feature, but can be helpful in some cases, as scaling the root for a voxel scene can become confusing when working with relative coordinates in code.

ClarkThyLord reviewed (commented):
Okay, I've taken a deeper look at everything, at it looks great 🙌 
I was able to make use of everything as expected, and only found a few small issues.
The reworked Vox importer looks and works great, I've only seen a few corner cases where it fails.
|MagicaVoxel|Godot|
|---|---|
|![Screenshot 2021-07-25 104816](https://user-images.githubusercontent.com/17155529/126931951-a352d284-26af-4124-9315-dca1b1a78f55.png)|![Screenshot 2021-07-25 104612](https://user-images.githubusercontent.com/17155529/126931946-7020c3c2-865a-45f8-827d-9965b5ee8c4c.png)|
|This is the model from within MagicaVoxel, as you can see it's two objects: One is the Knight with no rotation, and the other is the Axe with some z-rotation. The model is saved as a project, making it a extended vox file.|This is that project imported into Godot as a VoxelObject, as you can see everything looks good, however the Axe is set one more voxel to the right, which causes it to overwrite the Knight's right hand.|

I found this issue while testing the importer, something similar happens when you import the same project as a MagicaVoxelScene.

|MagicaVoxel|Godot|
|---|---|
|![Screenshot 2021-07-25 212352](https://user-images.githubusercontent.com/17155529/126932887-c31bf2ba-09a3-46f5-9453-cba44bbaaf81.png)|![Screenshot 2021-07-25 212332](https://user-images.githubusercontent.com/17155529/126932896-ef29551c-3852-4914-ab7a-53bc3a40568c.png)|
|Side view of the same project from within MagicaVoxel.|This is that project imported into Godot as a MagicaVoxelScene, the Axe is set one more voxel to the right and forward.|

ZIP for this MagicaVoxel scene: [chr_knight_and_axe.project.zip](https://github.com/ClarkThyLord/Voxel-Core/files/6875577/chr_knight_and_axe.project.zip)

Not sure if you know what might be causing this? I'm thinking it has to do with the way MagicaVoxel's positioning work, or maybe the center of these "objects" are not being calculated correctly somehow 🤔 

Other then that everything looks great, so I'd be willing to merge right now, unless you have any objections, thanks again for this great work! 🙌 

I've also gone ahead and done some refactoring and formatting to the code to make it consistent with the rest of the code base 👍

ElectronicBlueberry commented:
Good find on that edge case! It was caused by the x axis being rounded down with `floor()`, while it should be rounded up during import, because it is flipped from MagicaVoxel. This also caused the origin to shift on the knight model, which is now also fixed.

No objections to a merge from my side, I used the importer the past couple of days, and it seems stable.

ClarkThyLord reviewed (approved):
Nice, just tested the changes and it looks good 👌

I'll probably rework the Readers later to make them more consistent, because right now what they return isn't really well defined. I also need to redesign them to be able to accommodate for different structures and needs in the future, like that of "voxel scenes".

But that's a different topic for another time, for now I'll move forward with the merge, thanks again for sharing the great work! 🙌
